### PR TITLE
Feature/appointment regular scheduling

### PR DESCRIPTION
**Pull request by dimitrijevic12**

### CODE REVIEW

File: PatientWebApplication/HealthClinic.DAL/Model/Patient/DoctorAppointment.cs [hunk 1]
```diff
@@ -28,5 +28,18 @@ public DoctorAppointment(int id, TimeSpan time, string date, int patientId, int
...
+        {
+            var item = obj as DoctorAppointment;
+            if (item == null)
+            {
+                return false;
+            }
+            else
+            {
+                return this.Date.Equals(item.Date) && this.DoctorUserId == item.DoctorUserId && this.PatientUserId == item.PatientUserId && this.RoomId.Equals(item.RoomId);
+            }
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Model/Patient/DoctorAppointment.cs:41]: Else statement unnecessary.

Comment (trki293) [reply PatientWebApplication/HealthClinic.DAL/Model/Patient/DoctorAppointment.cs:41]: Use the method to check the equality of two strings, to check the date as well as to check the RoomId.

File: PatientWebApplication/HealthClinic.DAL/Model/Patient/DoctorAppointment.cs [hunk 2]
```diff
@@ -28,5 +28,18 @@ public DoctorAppointment(int id, TimeSpan time, string date, int patientId, int
 
         public DoctorAppointment() : base() { }
 
+        public override bool Equals(Object obj)
+        {
+            var item = obj as DoctorAppointment;
+            if (item == null)
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Model/Patient/DoctorAppointment.cs:34]: Use the method to check if the variable is null.

File: PatientWebApplication/HealthClinic.DAL/Service/DoctorService.cs [hunk 3]
```diff
@@ -31,24 +33,24 @@ public class DoctorService : AbstractUserService<DoctorUser>
...
-            doctorRepository = new DoctorRepository(path);          
-            operationRepository = new OperationRepository();
-            appointmentRepository = new AppointmentRepository();
-            employeesScheduleRepository = new EmployeesScheduleRepository(path5);
-
+            this._doctorRepository = doctorRepository;
+            this._operationRepository = operationRepository;
+            this._appointmentRepository = appointmentRepository;
+            this._employeesScheduleRepository = employeesScheduleRepository;
+            this.regularAppointmentService = new RegularAppointmentService(appointmentRepository, employeesScheduleRepository, this, new PatientsRepository(), new OperationService(operationRepository));
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/DoctorService.cs:42]: Need to delete 'this'.

File: PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs [hunk 4]
```diff
@@ -92,17 +101,81 @@ public DoctorAppointment RecommendAnAppointment(DoctorUser doctor, DateTime date
...
         }
 
+        public List<DoctorAppointment> GetAllAvailableAppointmentsForDate(string dateString, int doctorId, int patientId)
+        {
+            DateTime date = UtilityMethods.ParseDateInCorrectFormat(dateString);
+            List<DoctorAppointment> availableAppointments = new List<DoctorAppointment>();
+            List<TimeSpan> startTimesAppointments = GetAllStartTimes(CreateAppointmentSetForDate(date, doctorId, patientId).ToList());
+            List<Operation> operations = operationService.CreateOperationtSetForDate(date, doctorId, patientId).ToList();
+            Shift doctorShift = employeesScheduleService.getShiftForDoctorForSpecificDay(dateString, doctorService.GetByid(doctorId));
+            if(doctorShift == null)
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs:147]: Apply a method that checks whether the variable is null.

File: PatientWebApplication/HealthClinic.DAL/Utility/UtilityMethods.cs [hunk 5]
```diff
@@ -23,5 +24,16 @@ public static Boolean TryParseDateInCorrectFormat(String dateString)
...
+
+        public static Boolean CheckForSpecialty(DoctorUser doctor, string specialty)
+        {
+            return doctor.speciality.Equals(specialty);
+        }
+
+        public static Boolean TryParseTimeSpanInCorrectFormat(String timeSpan)
+        {
+            var time = new TimeSpan();
+            return TimeSpan.TryParseExact(timeSpan, "hh:\\mm\\:ss", CultureInfo.InvariantCulture, out time);
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Utility/UtilityMethods.cs:36]: Write in the form:
return TimeSpan.TryParseExact(timeSpan, "hh:\\mm\\:ss", CultureInfo.InvariantCulture, out new TimeSpan());

**Review by trki293 — CHANGES_REQUESTED** [review]
Correct the indicated places in the code, as well as the error that occurred during the build of the tests.

**Review by trki293 — APPROVED** [review]
PR is approved.

File: PatientWebApplication/HealthClinic.DAL/Service/DoctorService.cs [hunk 6]
```diff
@@ -224,33 +226,56 @@ public bool checkIfDoctorIsBusyForOperation(Operation operation, TimeSpan select
...
+        {
+            List<DoctorUser> specialists = new List<DoctorUser>();
+            foreach (DoctorUser doctor in GetAll())
+            {
+                if (UtilityMethods.CheckForSpecialty(doctor, specialty))
+                {
+                    specialists.Add(doctor);
+                }
+            }
+            return specialists;
```
Comment (Nemanja0505) [inline PatientWebApplication/HealthClinic.DAL/Service/DoctorService.cs:265]: Refactor code. Replace method GetDoctorsBySpecialty(string specialty)  with Where syntax ,for example : 
return GetAll().Where(doctor => UtilityMethods.CheckForSpecialty(doctor, specialty) )

File: PatientWebApplication/HealthClinic.DAL/Service/DoctorService.cs [hunk 7]
```diff
@@ -224,33 +226,56 @@ public bool checkIfDoctorIsBusyForOperation(Operation operation, TimeSpan select
...
+        {
+            List<DoctorUser> availableDoctors = new List<DoctorUser>();
+            foreach(DoctorUser doctor in GetDoctorsBySpecialty(specialty))
+            {
+                if(this.regularAppointmentService.GetAllAvailableAppointmentsForDate(date, doctor.id, patientId).Count != 0)
+                {
+                    availableDoctors.Add(doctor);
+                }
+            }
+            return availableDoctors;
```
Comment (Nemanja0505) [inline PatientWebApplication/HealthClinic.DAL/Service/DoctorService.cs:278]: Refactor code. Replace method GetAvailableDoctors(string specialty, string date, int patientId) with Where syntax ,for example :
return GetDoctorsBySpecialty(specialty).Where(doctor => this.regularAppointmentService.GetAllAvailableAppointmentsForDate(date, doctor.id, patientId).Count != 0 )

File: PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs [hunk 8]
```diff
@@ -197,5 +197,46 @@ private Boolean CheckIfOperation(String stringToCheck)
...
+        {
+            List<Operation> operations = new List<Operation>();
+            foreach (Operation operation in _operationRepository.GetOperationsForDoctor(doctorId))
+            {
+                if (date == UtilityMethods.ParseDateInCorrectFormat(operation.Date))
+                {
+                    operations.Add(operation);
+                }
+            }
+            return operations;
```
Comment (Nemanja0505) [inline PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs:210]: Refactor code. Replace method  GetAllOperationsByDateAndDoctor(DateTime date, int doctorId) with Where syntax ,for example :
return _operationRepository.GetOperationsForDoctor(doctorId).Where(operation => date == UtilityMethods.ParseDateInCorrectFormat(operation.Date) )

File: PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs [hunk 9]
```diff
@@ -197,5 +197,46 @@ private Boolean CheckIfOperation(String stringToCheck)
...
+        {
+            List<Operation> operations = new List<Operation>();
+            foreach (Operation operation in GetOperationsForPatient(patientId))
+            {
+                if (date == UtilityMethods.ParseDateInCorrectFormat(operation.Date))
+                {
+                    operations.Add(operation);
+                }
+            }
+            return operations;
```
Comment (Nemanja0505) [inline PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs:223]: Refactor code. Replace method GetAllOperationsByDateAndPatient(DateTime date, int patientId) with Where syntax ,for example :
return GetOperationsForPatient(patientId).Where(operation => date == UtilityMethods.ParseDateInCorrectFormat(operation.Date) )

**Review by Nemanja0505 — CHANGES_REQUESTED** [review]
PR is not approved refactor marked parts .

**Review by Nemanja0505 — APPROVED** [review]
Clean code is following and PR can be approved .

**Pull request merged**